### PR TITLE
Make bech32m the default for RPC, opt-in for GUI

### PR DESCRIPTION
**Pull request by Sjors**
**RPC**
This makes bech32m the default whenever possible. For legacy wallets the default remains bech32. Same for descriptor wallets without an active Taproot descriptor.

**GUI**
A new wallet setting allows users to opt-in to Taproot. If they opt in, the GUI bech32 checkbox now produces either bech32 or bech32m depending on the wallet. This avoids having to explain the difference.

Schermafbeelding 2021-10-20 om 20 13 17

**Considerations**
This lets users and developers experiment with taproot wallets. When merged before #22364, the user will have to manually import a taproot descriptor. After that PR, taproot descriptors will be present for all new wallets.

It is better to stick to bech32 addresses for the time being, because:
1. It is expected that it will take a while for enough wallets and exchanges to correctly handle sending to bech32m addresses
2. Taproot savings for a single signature wallet are not huge
3. Explaining the difference between bech32 and bech32m is not great UX

But we don't want to prevent users and developers who _do_ understand the difference to use Taproot. Without this PR they can only do this via the RPC (`getnewaddress some_label bech32m`). With this PR it's a simple checkbox.

Changing the RPC default to bech32m seems acceptable to me, since those users will have a better understanding of the wallet. Though I can see the case against it because it might be a breaking change, e.g. if someone has an automated setup that creates new wallets (existing wallets currently don't auto upgrade to taproot, so their behavior is not changed). 

**Suggested testing**
* call `getnewaddress` without type and `getnewaddress Test bech32` / `bech32m`
* create a receive address in the GUI with and without the bech32 checkbox
1. Legacy wallet
2. Default descriptor wallet (current does not have taproot descriptor)
3. Wallet with taproot descriptor:
   * build #22364 
   * create new descriptor wallet
   * build this PR again

### CODE REVIEW

**Comment by michaelfolkson** [comment]
My understanding from discussion at they July 30th Core [wallet meeting](https://gnusha.org/bitcoin-core-dev/2021-07-30.log) is that this will only be merged once the (super)majority of the ecosystem is able to make payments to bech32m addresses. So definitely post Taproot activation but uncertain at this stage when exactly.

Assuming I've understood above, Concept ACK. Thanks for the instructions for testing in the PR description, will do this testing at a later date.

**Review by ghost — APPROVED** [review]
tACK https://github.com/bitcoin/bitcoin/pull/22260/commits/2c2a8c0e29f3ccf9a4907c5870ccb41aa4fd1f42

https://bitcoin.stackexchange.com/a/108013/

**Comment by prayank23** [comment]
Not sure why I don't see QR in `bitcoin-qt`

image

File: src/qt/receivecoinsdialog.cpp [hunk 1]
```diff
@@ -144,12 +144,14 @@ void ReceiveCoinsDialog::on_receiveButton_clicked()
...
+
     if (ui->useBech32->isChecked()) {
-        address_type = OutputType::BECH32;
+        if (address_type != OutputType::BECH32 && address_type != OutputType::BECH32M) {
+            address_type = OutputType::BECH32;
+        }
     } else {
-        address_type = model->wallet().getDefaultAddressType();
-        if (address_type == OutputType::BECH32) {
+        if (address_type == OutputType::BECH32 ||address_type == OutputType::BECH32M) {
```
Comment (benthecarman) [inline src/qt/receivecoinsdialog.cpp:154]: spacing/formatting is weird here

**Comment by Sjors** [comment]
Rebased.

@prayank23 do you have libqrencode on your system? See https://github.com/bitcoin/bitcoin/blob/master/doc/build-unix.md#dependencies

**Comment by prayank23** [comment]
libqrencode was missing. I installed it with `sudo apt-get install qrencode`, still QR doesn't work. Its not related to this PR and maybe something wrong with this VM. I will try to fix this.

ACK https://github.com/bitcoin/bitcoin/commit/1ba76c2ce3446d0fd95f89bb91089c6664b29afc

**Comment by benthecarman** [comment]
> libqrencode was missing. I installed it with `sudo apt-get install qrencode`, still QR doesn't work. Its not related to this PR and maybe something wrong with this VM. I will try to fix this.

you probably need to reconfigure and rebuild

File: src/wallet/rpcdump.cpp [hunk 2]
```diff
@@ -1555,6 +1555,10 @@ static UniValue ProcessDescriptorImport(CWallet& wallet, const UniValue& data, c
                 // Taproot is not active, raise an error
                 throw JSONRPCError(RPC_WALLET_ERROR, "Cannot import tr() descriptor when Taproot is not active");
             }
+            // When importing an active taproot descriptor, change the default address type to bech32m
```
Comment (Deadyourface) [inline src/wallet/rpcdump.cpp:1558]: ### 

> `****`

**Comment by schildbach** [comment]
I wonder why this isn't already merged for Testnet and Signet (not Mainnet), both of which have Taproot activated already. It would make Taproot testing much easier.

**Pull request closed without merging**